File: sdk/npm/src/normalize.ts
import { Email, EmailAttachment } from './types';

/**
 * 将各提供商返回的原始邮件数据标准化为统一的 Email 格式
 *
 * 不同渠道的 API 返回字段名各不相同，此函数通过多字段候选策略
 * 将它们统一映射为标准的 Email 结构，保证 SDK 输出一致性。
 *
 * @param raw - 原始邮件数据（来自不同提供商的 API 响应）
 * @param recipientEmail - 收件人邮箱地址，当原始数据中无收件人字段时用作回退值
 * @returns 标准化的 Email 对象
 */
export function normalizeEmail(raw: any, recipientEmail: string = ''): Email {
  let text = normalizeText(raw);
  let html = normalizeHtml(raw);

  /*
   * 修正 text/html 错配：
   * 部分渠道将 HTML 内容放在 body/text 等字段中，
   * 导致 normalizeText 提取到 HTML 内容而 normalizeHtml 为空。
   * 检测 text 中是否包含 HTML 标签，如果是则移动到 html 字段。
   */
  if (text && !html && isHtmlContent(text)) {
    html = text;
    text = '';
  }

  return {
    id: normalizeId(raw),
    from: normalizeFrom(raw),
    to: normalizeTo(raw, recipientEmail),
    subject: normalizeSubject(raw),
    text,
    html,
    date: normalizeDate(raw),
    isRead: normalizeIsRead(raw),
    attachments: normalizeAttachments(raw.attachments),
  };
}

/**
 * 检测内容是否为 HTML
 * 通过检查是否包含常见的 HTML 标签来判断
 */
function isHtmlContent(content: string): boolean {
  const trimmed = content.trim().toLowerCase();
  return trimmed.startsWith('<!doctype html') ||
    trimmed.startsWith('<html') ||
    trimmed.startsWith('<body') ||
    (trimmed.includes('<div') && trimmed.includes('</div>')) ||
    (trimmed.includes('<table') && trimmed.includes('</table>')) ||
    (trimmed.includes('<p') && trimmed.includes('</p>') && trimmed.includes('<'));
}

/**
 * 提取邮件 ID
 * 候选字段: id, eid, _id, mailboxId, messageId, mail_id
 */
function normalizeId(raw: any): string {
  const id = raw.id ?? raw.eid ?? raw._id ?? raw.mailboxId ?? raw.messageId ?? raw.mail_id ?? '';
  return String(id);
}

/**
 * 提取发件人地址
 * 候选字段: from_address, address_from, from, messageFrom, sender
 */
function normalizeFrom(raw: any): string {
  return (
    raw.from_addr ||
    raw.from_address ||
    raw.fromAddress ||
    raw.mail_sender ||
    raw.sender ||
    raw.address_from ||
    raw.from_email ||
    raw.from ||
    raw.messageFrom ||
    ''
  );
}

/**
 * 提取收件人地址，无匹配字段时回退为 recipientEmail
 * 候选字段: to, to_address, name_to, email_address, address
 */
function normalizeTo(raw: any, recipientEmail: string): string {
  return raw.to || raw.to_address || raw.toAddress || raw.name_to || raw.email_address || raw.address || recipientEmail || '';
}

/**
 * 提取邮件主题
 * 候选字段: subject, e_subject
 */
function normalizeSubject(raw: any): string {
  return raw.subject || raw.e_subject || raw.mail_title || '';
}

/**
 * 提取纯文本内容
 * 候选字段: text, body, content, body_text, text_content
 */
function normalizeText(raw: any): string {
  return (
    raw.text ||
    raw.text_body ||
    raw.preview_text ||
    raw.mail_body_text ||
    raw.body ||
    raw.content ||
    raw.body_text ||
    raw.text_content ||
    raw.description ||
    ''
  );
}

/**
 * 提取 HTML 内容
 * 候选字段: html, html_content, body_html
 */
function normalizeHtml(raw: any): string {
  return raw.html || raw.html_body || raw.html_content || raw.body_html || raw.mail_body_html || '';
}

/**
 * 提取并统一日期格式为 ISO 8601
 * 候选字段: received_at, created_at, createdAt, date, timestamp, e_date
 * 其中 timestamp 为 Unix 秒级时间戳，需乘以 1000 转为毫秒
 */
function normalizeDate(raw: any): string {
  try {
    if (raw.received_at) return new Date(raw.received_at).toISOString();
    if (raw.receivedAt) return new Date(String(raw.receivedAt).replace(' ', 'T')).toISOString();
    if (raw.created_at) return new Date(raw.created_at).toISOString();
    if (raw.createdAt) return new Date(raw.createdAt).toISOString();
    if (raw.date) {
      if (typeof raw.date === 'number') return new Date(raw.date).toISOString();
      return new Date(raw.date).toISOString();
    }
    if (raw.timestamp) return new Date(raw.timestamp * 1000).toISOString();
    if (raw.e_date) return new Date(raw.e_date).toISOString();
  } catch {
    /* 日期解析失败，返回空字符串 */
  }
  return '';
}

/**
 * 提取已读状态
 * 候选字段: seen, read, isRead, is_read
 * 支持 boolean / number(0|1) / string('0'|'1') 多种类型
 */
function normalizeIsRead(raw: any): boolean {
  if (typeof raw.seen === 'boolean') return raw.seen;
  if (typeof raw.read === 'boolean') return raw.read;
  if (typeof raw.isRead === 'boolean') return raw.isRead;
  if (typeof raw.is_seen === 'number') return raw.is_seen === 1;
  if (typeof raw.is_seen === 'string') return raw.is_seen === '1';
  if (typeof raw.is_read === 'number') return raw.is_read === 1;
  if (typeof raw.is_read === 'string') return raw.is_read === '1';
  if (typeof raw.is_read === 'boolean') return raw.is_read;
  return false;
}

/**
 * 提取并标准化附件列表
 * 每个附件的字段也采用多候选策略映射
 */
function normalizeAttachments(attachments: any): EmailAttachment[] {
  if (!attachments || !Array.isArray(attachments)) return [];
  return attachments.map((a: any) => ({
    filename: a.filename || a.name || '',
    size: a.size || a.filesize || undefined,
    contentType: a.contentType || a.content_type || a.mimeType || a.mime_type || undefined,
    url: a.url || a.download_url || a.downloadUrl || undefined,
  }));
}

File: sdk/go/normalize.go
package tempemail

import (
	"encoding/json"
	"fmt"
	"math"
	"time"
)

/*
 * normalizeRawEmails 将原始 JSON 消息列表标准化为统一 Email 格式
 * 逐条解析 JSON 并调用 normalizeRawEmail 进行字段映射
 */
func normalizeRawEmails(rawMessages []json.RawMessage, recipientEmail string) ([]Email, error) {
	emails := make([]Email, 0, len(rawMessages))
	for _, raw := range rawMessages {
		var m map[string]interface{}
		if err := json.Unmarshal(raw, &m); err != nil {
			return nil, err
		}
		emails = append(emails, normalizeRawEmail(m, recipientEmail))
	}
	return emails, nil
}

/*
 * normalizeRawEmail 将原始 map 数据标准化为统一 Email 格式
 * 不同渠道的 API 返回字段名各不相同，通过多字段候选策略统一映射
 */
func normalizeRawEmail(raw map[string]interface{}, recipientEmail string) Email {
	return Email{
		ID:          normalizeID(raw),
		From:        normalizeFrom(raw),
		To:          normalizeTo(raw, recipientEmail),
		Subject:     normalizeSubject(raw),
		Text:        normalizeText(raw),
		HTML:        normalizeHTML(raw),
		Date:        normalizeDate(raw),
		IsRead:      normalizeIsRead(raw),
		Attachments: normalizeAttachments(raw),
	}
}

/*
 * getStr 从 map 中按优先级顺序提取字符串值
 * 依次尝试每个候选 key，返回第一个非空值，全部未命中时返回空字符串
 */
func getStr(m map[string]interface{}, keys ...string) string {
	for _, key := range keys {
		if v, ok := m[key]; ok && v != nil {
			switch val := v.(type) {
			case string:
				return val
			default:
				return fmt.Sprintf("%v", val)
			}
		}
	}
	return ""
}

/* normalizeID 提取邮件 ID，候选字段: id, eid, _id, mailboxId, messageId, mail_id */
func normalizeID(raw map[string]interface{}) string {
	return getStr(raw, "id", "eid", "_id", "mailboxId", "messageId", "mail_id")
}

/* normalizeFrom 提取发件人地址，候选字段: from_address, address_from, from, messageFrom, sender */
func normalizeFrom(raw map[string]interface{}) string {
	return getStr(raw, "from_address", "address_from", "from", "messageFrom", "sender")
}

/* normalizeTo 提取收件人地址，无匹配字段时回退为 recipientEmail */
func normalizeTo(raw map[string]interface{}, recipientEmail string) string {
	result := getStr(raw, "to", "to_address", "name_to", "email_address", "address")
	if result == "" {
		return recipientEmail
	}
	return result
}

/* normalizeSubject 提取邮件主题，候选字段: subject, e_subject */
func normalizeSubject(raw map[string]interface{}) string {
	return getStr(raw, "subject", "e_subject")
}

/* normalizeText 提取纯文本内容，候选字段: text, body, content, body_text, text_content */
func normalizeText(raw map[string]interface{}) string {
	return getStr(raw, "text", "body", "content", "body_text", "text_content")
}

/* normalizeHTML 提取 HTML 内容，候选字段: html, html_content, body_html */
func normalizeHTML(raw map[string]interface{}) string {
	return getStr(raw, "html", "html_content", "body_html")
}

/*
 * normalizeDate 提取并统一日期格式为 RFC3339
 * 候选字段: received_at, created_at, createdAt, date, timestamp, e_date
 * 支持字符串日期、秒级时间戳、毫秒级时间戳多种格式
 */
func normalizeDate(raw map[string]interface{}) string {
	/* 尝试字符串日期字段 */
	for _, key := range []string{"received_at", "created_at", "createdAt", "date"} {
		if v, ok := raw[key]; ok && v != nil {
			switch val := v.(type) {
			case string:
				if t, err := time.Parse(time.RFC3339, val); err == nil {
					return t.UTC().Format(time.RFC3339)
				}
				if t, err := time.Parse(time.RFC3339Nano, val); err == nil {
					return t.UTC().Format(time.RFC3339)
				}
				// 尝试直接返回字符串
				return val
			case float64:
				/* 大数字可能是毫秒时间戳，小数字为秒级时间戳 */
				if val > 1e12 {
					return time.UnixMilli(int64(val)).UTC().Format(time.RFC3339)
				}
				return time.Unix(int64(val), 0).UTC().Format(time.RFC3339)
			}
		}
	}

	/* 尝试数字时间戳字段，timestamp 为秒级，e_date 为毫秒级 */
	for _, key := range []string{"timestamp", "e_date"} {
		if v, ok := raw[key]; ok && v != nil {
			if num, ok := v.(float64); ok && num > 0 {
				if key == "timestamp" && num < 1e12 {
					return time.Unix(int64(num), 0).UTC().Format(time.RFC3339)
				}
				return time.UnixMilli(int64(num)).UTC().Format(time.RFC3339)
			}
		}
	}

	return ""
}

/*
 * normalizeIsRead 提取已读状态
 * 候选字段: seen, read, isRead, is_read
 * 支持 bool / float64(0|1) / string("0"|"1") 多种类型
 */
func normalizeIsRead(raw map[string]interface{}) bool {
	/* 布尔值 seen (mail.tm) */
	if v, ok := raw["seen"]; ok {
		if b, ok := v.(bool); ok {
			return b
		}
	}
	/* 布尔值 read */
	if v, ok := raw["read"]; ok {
		if b, ok := v.(bool); ok {
			return b
		}
	}
	/* 布尔值 isRead */
	if v, ok := raw["isRead"]; ok {
		if b, ok := v.(bool); ok {
			return b
		}
	}
	/* 数字或字符串 is_read */
	if v, ok := raw["is_read"]; ok {
		switch val := v.(type) {
		case float64:
			return int(math.Round(val)) != 0
		case bool:
			return val
		case string:
			return val == "1"
		}
	}
	return false
}

/*
 * normalizeAttachments 提取并标准化附件列表
 * 每个附件的字段也采用多候选策略映射
 */
func normalizeAttachments(raw map[string]interface{}) []EmailAttachment {
	v, ok := raw["attachments"]
	if !ok || v == nil {
		return []EmailAttachment{}
	}

	arr, ok := v.([]interface{})
	if !ok {
		return []EmailAttachment{}
	}

	attachments := make([]EmailAttachment, 0, len(arr))
	for _, item := range arr {
		m, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		att := EmailAttachment{
			Filename: getStr(m, "filename", "name"),
		}
		if size, ok := m["size"].(float64); ok {
			att.Size = int64(size)
		} else if size, ok := m["filesize"].(float64); ok {
			att.Size = int64(size)
		}
		att.ContentType = getStr(m, "contentType", "content_type", "mimeType", "mime_type")
		att.URL = getStr(m, "url", "download_url", "downloadUrl")
		attachments = append(attachments, att)
	}
	return attachments
}
